### PR TITLE
Feature/bulk import organisations

### PR DESCRIPTION
**Pull request by appsol**
### Bulk import of organisations

- Ref: https://trello.com/c/NUDw7odt
- Import and validate rows from .xls or .xlsx spreadsheet
- Returns total organisations imported or details of failing rows
- If any row fails no organisation is imported

### CODE REVIEW

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        $spreadsheetHandler->readHeaders();
+
+        $importedRows = 0;
+
+        DB::transaction(function () use ($spreadsheetHandler, &$importedRows) {
+            $rowBatch = [];
+            foreach ($spreadsheetHandler->readRows() as $row) {
+                $row['id'] = (string) Str::uuid();
+                $row['slug'] = Str::slug($row['name'] . ' ' . uniqid(), '-');
+                $rowBatch[] = $row;
```
Comment (matthew-inamdar) [inline app/Http/Controllers/Core/V1/Organisation/ImportController.php:196]: Can we add the `created_at` and `updated_at` timestamps?

```suggestion
                $row['id'] = (string) Str::uuid();
                $row['slug'] = Str::slug($row['name'] . ' ' . uniqid(), '-');
                $row['created_at'] = Date::now();
                $row['updated_at'] = Date::now();
                $rowBatch[] = $row;
```

File: app/Http/Requests/Organisation/ImportRequest.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'spreadsheet' => [
+                'required',
+                'regex:/^data:application\/vnd[a-z\-\.]+;base64,/',
```
Comment (matthew-inamdar) [inline app/Http/Requests/Organisation/ImportRequest.php:32]: This seems to fail when attempting to upload `.xls` file. It works fine with a `.xlsx` file.

Vue devtools whos the base64 encoded file as `data:application/octet-stream;base64,0M8R...` for `.xls`.

**Review by matthew-inamdar — CHANGES_REQUESTED** [review]
Looks good - I love the use of generators 👍 

Some comments left. Feel free to drop me a message if you want to go over any of them.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * Display the specified resource.
+     *
+     * @param \App\Http\Requests\Organisation\ImportRequest $request
+     * @throws Illuminate\Validation\ValidationException
+     * @return \Illuminate\Http\Response
+     */
+    public function __invoke(ImportRequest $request)
+    {
+        Log::info($request->input('spreadsheet'));
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:42]: Done (sorry).

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        $filePath = $this->storeBase64FileString($request->input('spreadsheet'), 'batch-upload');
+        Log::info($filePath);
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:44]: Done (sorry again).

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        return $rejectedRows;
+    }
+
+    /**
+     * Import the uploaded file contents
+     *
+     * @param String $filePath
+     * @return null
+     **/
+    public function importSpreadsheet(String $filePath)
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:181]: Done.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        return (new MimeTypes)->getExtensions($mime_type)[0] ?? null;
+    }
+
+    /**
+     * Validate the spreadsheet rows
+     *
+     * @param String $filePath
+     * @return Array
+     **/
+    public function validateSpreadsheet(String $filePath)
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:140]: Done.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    }
+
+    /**
+     * Store a Base 64 encoded data string
+     *
+     * @param string $file_data
+     * @param string $path
+     * @return String
+     **/
+    public function storeBase64FileString(string $file_data, string $path)
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:97]: Done.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * Store a binary file blob and update the models properties
+     *
+     * @param String $blob
+     * @param String $path
+     * @param String $mime_type
+     * @param String $ext
+     * @return String
+     **/
+    public function storeBinaryUpload(string $blob, string $path, $mime_type = null, $ext = null)
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:80]: Done.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        $rejectedRows = $this->validateSpreadsheet($filePath);
+        $importedRows = 0;
+
+        if (!count($rejectedRows)) {
+            $importedRows = $this->importSpreadsheet($filePath);
+        }
+
+        Storage::disk('local')->delete($filePath);
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:57]: Done.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php (hunk 1 above)
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php:196]: Done.

File: app/Http/Controllers/Core/V1/Organisation/ImportController.php [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+                if (count($rowBatch) === self::ROW_IMPORT_BATCH_SIZE) {
+                    DB::table('organisations')->insert($rowBatch);
+                    $importedRows += self::ROW_IMPORT_BATCH_SIZE;
+                    $rowBatch = [];
+                }
+            }
+
+            if (count($rowBatch) && count($rowBatch) !== self::ROW_IMPORT_BATCH_SIZE) {
+                DB::table('organisations')->insert($rowBatch);
```
[the remark this answers is not in the document]
Comment (appsol) [reply app/Http/Controllers/Core/V1/Organisation/ImportController.php]: I've added in another transaction that bulk creates the Organisation Admins. This should be fine as we are working on the theory that these are all new organisations. It doesn't appear to add appreciably to the processing time and I have included it in the transaction so if anything goes wrong nothing is committed to disk.

**Review by matthew-inamdar — APPROVED** [review]
LGTM 👍  Nice work on this!

**Pull request merged**